Add rvalue reference ctors for TError class.

dreamer-dead opened the pull request:
TError is a widely used class in Porto but it lacks move-ctors and operator=.
It is possible to optimize creating and returning error values from functions.

Also I fixed operator bool to be explicit and it helps to find some strange lines.
Please, take a look.

dreamer-dead commented on src/portod.cpp:
```diff
@@ -908,7 +908,7 @@ static int MasterMain(bool respawn) {
     std::shared_ptr<TEpollLoop> ELoop = std::make_shared<TEpollLoop>();
     TError error = ELoop->Create();
     if (error)
-        return error;
```
Here was conversion from bool to int, but explicitly return of EXIT_FAILURE will be better IMHO.

dreamer-dead commented on src/task.cpp:
```diff
@@ -725,7 +725,7 @@ TError TTask::ChildSetHostname() {
         string host = Env->Hostname + "\n";
         TError error = f.WriteStringNoAppend(host);
         if (error)
-            return TError(EError::Unknown, error, "write(/etc/hostname)");
```
And this is another interesting line.
Operator bool returns only true or false and the real errno value will be lost.

dreamer-dead commented on src/test/selftest.cpp:
```diff
@@ -4918,7 +4918,7 @@ static void TestCgroups(TPortoAPI &api) {
 
     TFolder qwerty(freezerCg);
     ExpectEq(qwerty.Exists(), true);
-    ExpectEq(qwerty.Remove(), false);
```
Remove() returns TError and there is a more suitable macro to check it.

dreamer-dead commented on src/util/error.cpp:
```diff
@@ -67,32 +105,28 @@ bool TError::Deserialize(int fd, TError &error) {
-    delete[] buf;
-
-    error = TError(err, desc, errno);
+    error = TError(err, std::move(desc), errno);
```
This function looks strange - it returns true even if it fails to read a TError value from file and assign to result value errno code.
If reading error appears, it will be impossible to if it is a deserialized error or it is a generated error for read operation.

dreamer-dead commented on src/util/error.cpp:
```diff
@@ -8,14 +8,53 @@ extern "C" {
 }
 
-TError::TError(EError e, int eno, std::string description) :
+TError::TError(EError e, int eno, const std::string &description) :
```
TError objects used to return results from many functions, and every object holds a string, even if it describes a errno.
How do you thinks, can we avoid storing description in every object?

antmat commented on src/util/error.cpp:
```diff
@@ -8,14 +8,53 @@ extern "C" {
+    Error(e), Description(std::move(description)), Errno(eno) {
+}
+
+TError::TError(const TError &other) :
```
Why do you need explicit copy and move constructors while compiler would generate them implicitly just in the way they are implemented here? If you do want to explicitly define a ctor you can use something like
'TError(TError&&) = default' in header file

dreamer-dead replied:
@antmat , good point, thanks.
I'll fix it.

rgushchin commented:
It was my comment)

dreamer-dead commented:
@rgushchin , I have rebased on the lastest master and removed explicit bodies of move ctors and operators.

Are you sure that your comment was for this PR? Maybe it was for #4 ?
I didn't add explicit keyword to constructors here =)
OK than, I'll do!

rgushchin commented:
Merged.
Thank you!
